### PR TITLE
Code Insights: Move insights to the enterprise version

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/24201

## Context

This PR moves code insights codebase from `./src/insights/` to `./src/enterprise/insights`. We already moved some parts from the code insight codebase to simplify this PR codebase migration. https://github.com/sourcegraph/sourcegraph/pull/24677. In this PR we do two things
- Most obvious and important thing we move the codebase 
- We introduce a new system for the extension view rendering for the OSS version and for the enterprise version.

### Extension views section component

At the moment we have code insights not only on the dashboard page but we render them on the directory and home (search) page. Historically at some point, we used the Extension API for fetching data for code insights. For that, we had introduced some extension API methods `getHomepageViews`, `getDirectoryView`, ..etc. At the moment we don't use this API for code insights since we had moved our data fetching from Extension API to the main app (PR https://github.com/sourcegraph/sourcegraph/pull/24559). But we still want to support these extension views. But we have code insights only in the enterprise version. In order to render extension like views on in OSS version and render extension like views with Code insights view in the enterprise version, this PR introduces component separation. You can find this system in `/src/insights/sections/` components.

### How to test
1. Check the OSS version (directory and homepage) of the sourcegraph. You should be able to see the only extensions like views like the Codecov or Snyk extensions views.
2. Check enterprise version (directory and homepage) you have to see extension and code insights views. 
3. Check that only in enterprise version you can see create code insight from the search page (1-click creation flow)

### Known issues/Work in progress
1. We should update Code Insights doc code links since this PR moves code insight codebase.
2. ~~We still have to fix enterprise imports~~
3. ~~Hide insight navigation item in OSS version~~

### CODE REVIEW

File: client/web/src/OpenSourceWebApp.tsx [hunk 1]
```diff
@@ -46,6 +47,7 @@ export const OpenSourceWebApp: React.FunctionComponent = () => (
         repoSettingsSidebarGroups={repoSettingsSideBarGroups}
         routes={routes}
         keyboardShortcuts={KEYBOARD_SHORTCUTS}
+        extensionViews={ExtensionViewsSection}
```
Comment (vovakulikov) [inline client/web/src/OpenSourceWebApp.tsx:50]: Pass OSS version of extension view section component

File: client/web/src/enterprise/insights/sections/ExtensionViewsDirectorySection.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    return (
+        <ViewGrid viewIds={allViewIds} telemetryService={props.telemetryService} className={className}>
+            {/* Render extension views for the directory page */}
+            {extensionViews.map(view => (
+                <StaticView key={view.id} view={view} telemetryService={props.telemetryService} />
+            ))}
+
+            {/* Render all code insights with proper directory page context */}
+            {directoryPageContext
```
Comment (vovakulikov) [inline client/web/src/enterprise/insights/sections/ExtensionViewsDirectorySection.tsx:99]: This is enterprise version of directory view section so we render extension like views and code insights

File: client/web/src/insights/sections/ExtensionViewsDirectorySection.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    return (
+        <ViewGrid
+            viewIds={extensionViews.map(view => view.id)}
+            telemetryService={props.telemetryService}
+            className={className}
+        >
+            {/* Render extension views for the directory page */}
+            {extensionViews.map(view => (
```
Comment (vovakulikov) [inline client/web/src/insights/sections/ExtensionViewsDirectorySection.tsx:72]: OSS version of directory page extension view section so renders only extension-like views.

File: client/web/src/OpenSourceWebApp.tsx [hunk 4]
```diff
@@ -5,6 +5,7 @@ import { extensionAreaRoutes } from './extensions/extension/routes'
 import { extensionsAreaHeaderActionButtons } from './extensions/extensionsAreaHeaderActionButtons'
 import { extensionsAreaRoutes } from './extensions/routes'
 import './SourcegraphWebApp.scss'
+import { ExtensionViewsSection } from './insights/sections/EtenstionViewsSection'
```
Comment (eseliger) [inline client/web/src/OpenSourceWebApp.tsx:8]:
```suggestion
import { ExtensionViewsSection } from './insights/sections/ExtensionViewsSection'
```

File: client/web/src/enterprise/insights/components/form/form-input/FormInput.tsx [hunk 5]
```diff
@@ -3,7 +3,7 @@ import React, { useRef, forwardRef, InputHTMLAttributes, ReactNode } from 'react
 import { useMergeRefs } from 'use-callback-ref'
 
 import { LoaderInput } from '@sourcegraph/branded/src/components/LoaderInput'
-import { useAutoFocus } from '@sourcegraph/wildcard'
+import { useAutoFocus } from '@sourcegraph/wildcard/src'
```
Comment (eseliger) [inline client/web/src/enterprise/insights/components/form/form-input/FormInput.tsx:6]: that seems odd

File: client/web/src/insights/analytics.ts [hunk 6]
```diff
@@ -6,15 +6,26 @@ import {
...
 import { TelemetryService } from '@sourcegraph/shared/src/telemetry/telemetryService'
 
-import { Settings } from '../../schema/settings.schema'
+import { Settings } from '../schema/settings.schema'
 
-import { INSIGHTS_ALL_REPOS_SETTINGS_KEY, InsightTypePrefix } from './types'
-import { isLangStatsdInsightId } from './types/insight/lang-stat-insight'
-import { isSearchBasedInsightId, SearchBasedExtensionInsightSettings } from './types/insight/search-insight'
-import { isGlobalSubject, isOrganizationSubject } from './types/subjects'
+const BACKEND_INSIHGTS_SETTINGS_KEY = 'insights.allrepos'
```
Comment (eseliger) [inline client/web/src/insights/analytics.ts:17]:
```suggestion
const BACKEND_INSIGHTS_SETTINGS_KEY = 'insights.allrepos'
```

**Review by eseliger — APPROVED** [review]
Great work! High level LGTM 🚀

File: client/web/src/enterprise/insights/components/form/form-input/FormInput.tsx (hunk 5 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/components/form/form-input/FormInput.tsx:6]: IDE fixed that for me. Apparently, IDE did it wrong :)

File: CHANGELOG.md [hunk 7]
```diff
@@ -17,6 +17,11 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Changed
 
+- `allowGroupsPermissionsSync` in the GitHub authorization provider is now required to enable the experimental GitHub teams and organization permissions caching. [#24561](https://github.com/sourcegraph/sourcegraph/pull/24561)
+- GitHub external code hosts now validate if a corresponding authorization provider is set, and emits a warning if not. [#24526](https://github.com/sourcegraph/sourcegraph/pull/24526)
+- Sourcegraph is now built with Go 1.17. [#24566](https://github.com/sourcegraph/sourcegraph/pull/24566)
```
Comment (unclejustin) [inline CHANGELOG.md:22]: Weird, are these supposed to be in this PR?

**Review by unclejustin — COMMENTED** [review]
Either this LGTM or you have simply overwhelmed me with changes. Either way APPROVED!

🌮

File: CHANGELOG.md (hunk 7 above)
Comment (vovakulikov) [reply CHANGELOG.md:22]: It's interesting. I can see that these three lines in this section have been added some time ago but for some reason, before this PR we didn't have them on main. 

@bobheadxi @camdencheek can you confirm that these changelogs are needed in this section, or it's my/git mistake?

Comment (vovakulikov) [reply CHANGELOG.md:22]: I see that with patch release 3.31.2 these lines had been removed 

https://github.com/sourcegraph/sourcegraph/pull/24753/files

Comment (camdencheek) [reply CHANGELOG.md:22]: Yeah, the "upgrade to 1.17" changelog should definitely be there. Looks like removing them for the patch release was a mistake

**Pull request merged**